riscv/csrs.cc

```diff
@@ -538,17 +542,46 @@ reg_t rv32_low_csr_t::written_value() const noexcept {
   return orig->written_value() & 0xffffffffU;
 }
 
+// implement class rv64_rv32_low_csr_t
+rv64_rv32_low_csr_t::rv64_rv32_low_csr_t(processor_t* const proc, const reg_t addr,
+                                         csr_t_p orig, priv_mode_t prv):
+  csr_t(proc, addr),
+  orig(orig),
+  low_half(std::make_shared<rv32_low_csr_t>(proc, addr, orig)),
+  prv(prv) {
+};
+
+reg_t rv64_rv32_low_csr_t::read() const noexcept {
+  unsigned xlen = proc->get_xlen(prv);
+  return xlen == 32 ? low_half->read() : orig->read();
+}
+
+void rv64_rv32_low_csr_t::verify_permissions(insn_t insn, bool write) const {
+  orig->verify_permissions(insn, write);
+}
+
+bool rv64_rv32_low_csr_t::unlogged_write(const reg_t val) noexcept {
+  unsigned xlen = proc->get_xlen(prv);
+  if (xlen == 32)
+    return low_half->unlogged_write(val);
+  else
+    return orig->unlogged_write(val);
+}
```

Comment:
This doesn't sound right. What about a CSR with S-mode privilege, such as `stimecmp`, which can be read from either M-mode or S-mode? If mxlen==64 and sxlen==32 then `csrw` should behave differently when executed from M-mode vs S-mode.

It appears to me that with this implementation, it will write only 32-bits even when run from M-mode where mxlen=64.

Reply:
Yes. From spec, stimecmp is SXLEN-bits length. So I think  it will only read/write 32-bits if sxlen = 32, no matter mxlen is 32 or 64.